torch/autograd/graph.py

```diff
@@ -144,42 +155,48 @@ def register_prehook(self, fn: Callable[..., Any]) -> RemovableHandle:
             >>> print(a.grad)
             tensor([1., 1., 1.])
         """
-        ...
+        raise NotImplementedError
 
     @classmethod
-    def __subclasshook__(cls, C):
-        if cls is Node:
-            if (
-                C is not None and C is getattr(torch._C._functions, C.__name__, None)
-            ) or issubclass(C, torch.autograd.function.BackwardCFunction):
-                return True
+    def __subclasshook__(cls, subclass: type) -> bool:
+        if cls is Node and (
+            (
+                subclass is not None
+                and subclass is getattr(torch._C._functions, subclass.__name__, None)
+            )
+            or issubclass(subclass, torch.autograd.function.BackwardCFunction)
+        ):
+            return True
         return NotImplemented
 
 
-def _get_grad_fn_or_grad_acc(t):
+def _get_grad_fn_or_grad_acc(t: torch.Tensor) -> Node:
     if t.requires_grad and t.grad_fn is None:
-        return t.view_as(t).grad_fn.next_functions[0][0]
-    else:
-        return t.grad_fn
+        return t.view_as(t).grad_fn.next_functions[0][0]  # type: ignore[union-attr,return-value]
+    return t.grad_fn  # type: ignore[return-value]
```

Comment:
Is this complaining that we need to annotate this return with `Optional[Node]` instead?

Reply:
This function is a private API and is only used in this file. All its usages in this file always return `Node`.

Reply:
What if I call register_multi_grad_hook with tensors that do not require grad

Reply:
`torch._C._will_engine_execute_node(grad_fn)` will raise `RuntimeError` if `grad_fn is None`.

https://github.com/pytorch/pytorch/blob/3fec0efd344b56235d118fd78527fe3896f4f2d8/torch/csrc/Module.cpp#L1159-L1167

`arg` should be a callable.

Reply:
That happens later though, when this function has already returned None right? If we want to say that this function should return Node (when used properly), let's add an assertion to enforce that? This would make things more consistent in case another usage in this file uses this API.

Reply:
Updated. I add an assertion.